### ISSUE TITLE
Unexpected application/vnd.docker.raw-stream frame type [72, 84, 84, 80, 47, 49, 46, 49]

### ISSUE THREAD

**Issue by Blogshot** [jenkinsci/docker-plugin#808]
Hello there!

My infrastructure is as follows:
`Jenkins (Docker Host) <-- Docker API --> Builder (Docker-Host)`

Jenkins is dockerized and has access to it's own docker host via Docker-CLI. There is also a remote host. Both are configured as cloud machines and pretty much work without complaint.

~~However, once every few days any pipeline step which invokes a new container might fail~~ My pipeline will reproducibly fail with something like this:
`java.io.IOException: Unexpected application/vnd.docker.raw-stream frame type [B@4b38e054`. 

There isn't really any stacktrace since the issue happens prior to everything else. There aren't any network issues that I know of and this can happen throughout the day, meaning there is no backup or such.

Details:
 - docker-plugin version: 1.2.0
 - jenkins version: 2.245
 - docker engine version: 19.03.6
 - Our images are based on `jenkins/slave`, connected via `Attach Docker container`

The pipeline is setup like this:

```
 Start     Fork #1                  Fork #2      End
   o ───┬─ Forked 1 ─┬─ Publish ─┬─ Forked 4 ─┬── o
        ├─ Forked 2 ─┤           ├─ Forked 5 ─┤
        └─ Forked 3 ─┘           └─ Forked 6 ─┘
```
The crash happens at stage "Publish".

Declarative pipeline:
```
pipeline {
  agent { label 'docker' }

  stages {
    stage('Fork #1') {
      parallel {
        stage('Forked 1') {
          agent { label 'docker' }
          stages { ... }
        }
        stage('Forked 2') {
          agent { label 'docker' }
          stages { ... }
        }
        stage('Forked 3') {
          agent { label 'docker' }
          stages { ... }
        }
      }
    }

    stage('Publish') {
      steps { ...}
    }

    stage('Fork #2') {
      parallel {
        stage('Forked 4') { ... }
        stage('Forked 5') { ... }
        stage('Forked 6') { ... }
      }
    }
  }
}
```

My current guess is that after the Fork, the Pipeline tries to return to the original Container it originated from. If this container doesn't exist anymore, the pipeline will consequently fail.

**Comment by Blogshot** [comment]
I did a bit of trial and error.

The following change to my layout works just fine:
```
 Start              Fork #1                  Fork #2      End
   o ─ Forked 1 ─┬─ Forked 2 ─┬─ Publish ─┬─ Forked 4 ─┬── o
                 └─ Forked 3 ─┘           ├─ Forked 5 ─┤
                                          └─ Forked 6 ─┘
```
Additionally, I've configured my **Idle Timeout** to 120 minutes instead of 10 minutes (default value). Still, the execution failed.

At this point, I've no idea on how to progress. For the time being, I changed my pipeline to the layout above to have it in a working state at least.

**Comment by pjdarton** [comment]
FYI the error you're seeing isn't caused by network issues - it's not a "connection drop" caused by packet loss etc.
It's coming from [this line](https://github.com/jenkinsci/docker-plugin/blob/docker-plugin-1.2.0/src/main/java/io/jenkins/docker/client/DockerMultiplexedInputStream.java#L70) which is complaining that it got the _wrong_ kind of data from the docker daemon.
Unfortunately, there's a bug in the code that reports that error which is why it's saying `[B@4b38e054` instead of something more meaningful.  That reporting bug is [fixed](https://github.com/jenkinsci/docker-plugin/commit/05c071775e9bf50a132c252027a1add705fca987) in the latest (non-released) version of the code (you can build it yourself or download a copy of the `.hpi` file from [here](https://ci.jenkins.io/job/Plugins/job/docker-plugin/job/master/203/artifact/io/jenkins/docker/docker-plugin/1.2.1-rc942.2c4503306727/)).

So, the _latest_ code will improve the error so it'll tell you more, but it won't make the problem go away; the problem is you're getting a bad response from the docker daemon sometimes, and _that's_ the reason it's failing.
My guess is that your docker daemon may be "being too clever" for the docker-plugin code - I'd guess that there's been a breaking change to the docker API since this code was written.  The docker plugin code expects [this api](https://docs.docker.com/engine/api/v1.32/#operation/ContainerAttach) and what you're seeing is a complaint from the docker-plugin code when it doesn't get that.

I'd suggest that you try the latest plugin code and see what the error changes to - that'll at least tell you what's going wrong.
Once you know what the docker-plugin is complaining about, hopefully that'll give you a clue as to what's happened.

**Comment by Blogshot** [comment]
Thank you very much for your insights!

I've replaced the files inside `/var/jenkins_home/plugins/docker-plugin/` with the contents of the `.hpi` you linked and restarted the Jenkins container. The plugin is correctly shown as "1.2.1-rc942.2c4503306727".

Then, I started the pipeline again in the original layout. The correct error message is:
`java.io.IOException: Unexpected application/vnd.docker.raw-stream frame type [72, 84, 84, 80, 47, 49, 46, 49]`

**Comment by pjdarton** [comment]
FYI you don't have to do file replacement; you can upload a plugin via the Jenkins UI (plugins, advanced tab); direct filesystem changes will also work (as you've shown) but it's "cleaner" to get Jenkins to do the filesystem stuff 😉 

So, `[72, 84, 84, 80, 47, 49, 46, 49]` is the 8 bytes that the docker-plugin received in response to its request.
That's `['H', 'T', 'T', 'P', '/', '1', '.', '1']`
That is _not_ the 8-byte header that this code was expecting...

...so, going backwards a bit, we got to this point from [here](https://github.com/jenkinsci/docker-plugin/blob/docker-plugin-1.2.0/src/main/java/io/jenkins/docker/connector/DockerComputerAttachConnector.java#L139) where we've sent a request to the docker-daemon asking it to attach to the container, we read its "HTTP/1.1 101 ..." response (and explode if we don't get that, so we're obviously getting that bit) and then read lines until we get a blank one (thus skipping over the rest of the HTTP response's headers) before we transfer control over to the code responsible for the exception you saw.

i.e. sometime _after_ the docker-daemon issued its "HTTP/1.1 101 ..." response + headers, we're getting what looks like another HTTP response ... and that's not something we expect; it's not something a well behaved docker-daemon should be doing either.

I don't know if it "worked for a while before it broke" or if it got this unwanted text immediately after the text it was expecting, but that's something you might be able to figure out from detailed logs at your end - if the container was able to do _some_ work before it disconnected then that tells us it "worked for a while before it broke".

I don't know why your docker-daemon is doing this ... or even if it _is_ the docker-daemon (there could be some as-yet-unidentified man-in-the-middle, e.g. a proxy), but you're going to need to track things down at your end to find this out.

If your docker-daemon is on another computer, you could use something like Wireshark (or just tcpdump+tcptrace) to log the traffic between Jenkins and the docker-daemon.
If your docker-daemon is on the same computer, you may be able to use strace (heavily filtered) to achieve a similar effect.
Regardless, it's probably a good idea to check netstat etc to verify that the connection from Jenkins to the docker-daemon is going direct and isn't going via a proxy (as, if it is, this may be the fault of the proxy).

TL;DR: You need to log the traffic between Jenkins and the docker-daemon, trace its path 100%, and figure out WTF is putting a second HTTP header into that stream.

**Comment by pjdarton** [comment]
Closing due to lack of activity; feel free to re-open if further information is forthcoming.
